www/content/essays/web-security-basics-with-htmx.md

```diff
@@ -0,0 +1,322 @@
+```
+
+But this is not:
+
+```html
+<button hx-get="https://google.com/search?q=events">Search events</button>
+```
+
+The reason for this is simple: htmx inserts the response from that route directly into the user's page. If the response has a malicious `<script>` inside it, that script can steal the user's data. When you don't control the route, you cannot guarantee that whoever does control the route won't add a malicious script.
+
+Fortunately, this is a very easy rule to follow. Hypermedia APIs (i.e. HTML) are [specific to the layout of your application](https://htmx.org/essays/hypermedia-apis-vs-data-apis/), so there is almost never any reason you'd *want* to insert someone else's HTML into your page. All you have to do is make sure you only call your own routes.
+
+Though it's not quite as popular these days, a common SPA pattern was to separate the frontend and backend into different repositories, and sometimes even to serve them from different URLs. This would require using absolute URLs in the frontend, and often, [disabling CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS). With htmx (and, to be fair, modern React with NextJS) this is an anti-pattern.
+
+Instead, you simply serve your HTML frontend from the same server (or at least the same domain) as your backend, and everything else falls into place: you can use relative URLs, you'll never have trouble with CORS, and you'll never call anyone else's backend.
+
+htmx executes HTML; HTML is code; never execute untrusted code.
+
+### Always use an auto-escaping template engine
+
+When you send HTML to the user, all dynamic content must be escaped. Use a template engine to construct your responses, and make sure that auto-escaping is on.
+
+Fortunately, all template engines support escaping HTML, and most of them enable it by default (Jinja is a notable exception). Below are just a few examples.
+
+| Language | Template Engine | Escapes HTML by default? |
+| ---- | ---- | ---- |
+| JavaScript | Nunjucks | Yes |
+| JavaScript | EJS | Yes, with `<%= %>` |
+| JavaScript | Handlebars | Yes, with `{{ }}` |
+| Python | Jinja | **No** |
+| Ruby | ERB | Yes, with `<%= %>` |
+| PHP | Blade | Yes |
+| Go | html/template | Yes |
+| Java | Thymeleaf | Yes |
+| Rust | Tera | Yes |
+
+The kind of vulnerability this prevents is often called a Cross-Site Scripting (XSS) attack, a term that is [broadly used](https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html#introduction) to mean the injection of any unexpected content into your webpage. Typically, an attacker uses your APIs to store malicious code in your database, which you then serve to your other users who request that info.
+
+For example, let's say you're building a dating site, and it lets users share a little bio about themselves. You'd render that bio like this, with `{{ user.bio }}` being the bio stored in the database:
+
+```html
+<p class=bio>
```

Comment:
These are all legal and fine

Reply:
True, but (1) it's inconsistent because other places in the article are using quotes and (2) this is a security-focused article and specifically recommends using quotes, so why not get into the habit everywhere?

Reply:
You're right, I was just going to come back and say that

Reply:
I stand by my no-quoting style!!! But it's not appropriate for this article specifically